Implement /reviewers/browse/ API.

EnTeQuAk opened the pull request:
Fixes #10322 

Contains a few very small changes along the way to simplify #9498

EnTeQuAk commented on src/olympia/reviewers/serializers.py:
```diff
@@ -1,11 +1,184 @@
+                  'release_notes', 'reviewed', 'url', 'version')
+
+
+class AddonFileBrowseSerializer(serializers.ModelSerializer):
```
I did and I think none of the arguments from the beginning are true anymore. Fixed in 2730656505, thanks for the reminder!

EnTeQuAk commented:
> The thing I don't like about all this is that it's a little weird to have file ids as the main identifier. We didn't have a file API until now because everything is based on either an add-on or version. The only reason we kinda need to expose a file API instead of a version API is for compatibility with the existing file viewer...

True, initially I thought that compatibility with the existing file-viewer and thus legacy add-ons and add-ons that have multiple platforms is a must-have but thinking about it it doesn't have to be. If we can get rid of add-ons that have multiple platforms (which aren't that many, at least add-ons that have different files for a platform) we can simply act on add-on version in the API rather than files directly.

Does that sound better?

The thing is, and that's where this is coming from, I tried to very early on rewrite the current file-viewer to use the serializer and thus the git-storage as a backend instead of extracting on its own. To still be able to do that we could expose the API via a version and internally forward `version.current_file` (`all_files[0]` to the serializer), does that sound okay?

EnTeQuAk commented on src/olympia/reviewers/tests/test_serializers.py:
```diff
@@ -0,0 +1,219 @@
+
+        filename = os.path.join(settings.TMP_PATH, 'test_isbinary')
+        for txt in ['#!/usr/bin/python', '#python', u'\0x2']:
+            open(filename, 'w').write(txt)
```
The problem with copy&paste :see_no_evil:… you become immediately the author of something and responsible :laughing: 

I don't have anything against it except that those are crazy junk files which is why I'd hesitate to do that… I'll change that, I feel like it'll be much better.

EnTeQuAk replied:
Fixed in 99fa29976c, luckily the `is_binary` function in the serializer now accepts the actual blob content which removes the need to write any files into the repo or on disk.

diox commented:
Yeah I like that. I'm not sure that we really care that much about porting the existing file viewer to the new API anyway...

EnTeQuAk commented on docs/topics/api/reviewers.rst:
```diff
@@ -76,9 +76,41 @@ add-on.
+
+------
+Browse
+------
```
0c3db23542

EnTeQuAk commented on src/olympia/reviewers/views.py:
```diff
@@ -1234,3 +1239,48 @@ def flags(self, request, **kwargs):
+        # If the instance is marked as deleted and the client is not allowed to
+        # see deleted instances, we want to return a 404, behaving as if it
+        # does not exist.
+        has_permission = (
```
Fixed in 944ce1a8b3, I went the other way around and moved the whole check into the separate variable which feels nicer I think.

EnTeQuAk commented:
@diox I rewrote the whole thing to be based off `version` now instead of `file` directly. I think it does feel better this way and leaves more room to extend the details later if needed, e.g to integrate some parts/actions of the actual reviewer tools or merge them.

Updated and extended docs, tests etc.

EnTeQuAk commented on src/olympia/lib/git.py:
```diff
@@ -18,6 +19,8 @@
 log = olympia.core.logger.getLogger('z.git_storage')
 
+# A mixture of Blob and TreeEntry
+TreeEntryWrapper = namedtuple('Entry', 'tree_entry, path, blob')
```
btw, I'm not happy with that name and open to suggestions. This thing wraps [TreeEntry](https://www.pygit2.org/objects.html#tree-entries) and extends it with some metadata. I wonder if I should just name it `TreeEntry` or just `Entry`?

diox replied:
I'm fine with it like this tbh

diox commented on src/olympia/reviewers/views.py:
```diff
@@ -1234,3 +1238,47 @@ def flags(self, request, **kwargs):
+        # If the instance is marked as deleted and the client is not allowed to
+        # see deleted instances, we want to return a 404, behaving as if it
+        # does not exist.
+        can_see_deleted_addons = (
```
By setting this in a variable early, you're making the query every time, even for non-deleted add-ons. Since it's just a single call I think it doesn't need to be a variable and could just be done in the condition below, where it'd be guarded by the `obj.deleted` part of the condition.
